### DISCUSSION TITLE
STARTTLS and SSL not working, TLS works

### DISCUSSION THREAD

**Discussion by david-brombacher** [postalserver/postal#2830] [Help with using Postal] [answered]
I need some help getting STARTTLS and SSL working - thank you all in advance
Only TLS is working, as log shows below.

I have made a new postal installation some weeks ago, version 2.1.4 on Ubuntu 22 TLS. I took over the SQL-Databases from the previous server running 1.x
After i got that basically working, i upgraded to the current release 2.3.1.
Caddy is running in a docker container (as suggested by the installation instructions)

For SSL/TLS/STARTTLS i did the following:

`ln /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.my-domain.com/postal.my-domain.com.crt /opt/postal/config/postal.my-domain.cert`
`ln /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.my-domain.com/postal.my-domain.com.key /opt/postal/config/postal.my-domain.key`
`chmod 644 /opt/postal/config/postal.my-domain.com*`

This covers the change of the extension for the certificate from .crt to .cert
After i upgraded to 2.3.1. i deleted the certificates to get new ones. I believe there is no need to go into caddy to make sure its RSA certificates, because that should be fixed in this version of postal.
To make sure, that the link of the files are not leading to the error, i also tried with copy. Services and even server restart after changes.

In the postal.yml:

smtp_server:
  port: 25
  tls_enabled: true
  tls_certificate_path: /config/postal.my-domain.com.cert
  tls_private_key_path: /config/postal.my-domain.com.key

I did postal stop/start and reboots afterwards, but the situation is:

Result from [https://www.gmass.co/smtp-test](https://www.gmass.co/smtp-test)

TLS is working:
```
Connected to smtp://postal.my-domain.com:25/?starttls=always
<< 220 postal.my-domain.com ESMTP Postal/AF6SGB
>> EHLO [172.31.11.248]
<< 250-My capabilities are
<< 250-STARTTLS
<< 250 AUTH CRAM-MD5 PLAIN LOGIN
>> STARTTLS
<< 220 Ready to start TLS
>> EHLO [172.31.11.248]
<< 250-My capabilities are
<< 250 AUTH CRAM-MD5 PLAIN LOGIN
>> AUTH CRAM-MD5
<< 334 PDJmMGVjODMxZWRjY2M5YmFhNTg5QHBvc3RhbC5kdmlzaW9uLm9yZz4=
>> YWRlbGJyZWNodC9zZXZkZXNrIDlmNDZiYzViYmE4ZDA1MDZiMDI4YzUzMTljNzY5NTIw
<< 235 Granted for postal-mail-sever/credential
```

SSL is not:

`ERROR: An error occurred while attempting to establish an SSL or TLS connection.`

Devices (printers with scan-to-mail) with STARTTLS do not work - limited log capabilities there.

This is the result from "openssl s_client -starttls smtp -crlf -connect postal.my-domain.com:25" (executed on a different server):

```
CONNECTED(00000003)
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
verify return:1
depth=0 CN = postal.my-domain.com
verify return:1
---
Certificate chain
 0 s:CN = postal.my-domain.com
   i:C = US, O = Let's Encrypt, CN = R3
   a:PKEY: id-ecPublicKey, 256 (bit); sigalg: RSA-SHA256
   v:NotBefore: Feb 29 10:50:58 2024 GMT; NotAfter: May 29 10:50:57 2024 GMT
 1 s:C = US, O = Let's Encrypt, CN = R3
   i:C = US, O = Internet Security Research Group, CN = ISRG Root X1
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Sep  4 00:00:00 2020 GMT; NotAfter: Sep 15 16:00:00 2025 GMT
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIEJTCCAw2gAwIBAgISBGBlc3QdN7kcgppeNrA6IEWRMA0GCSqGSIb3DQEBCwUA
MDIxCzAJBgNVBAYTAlVTMRYwFAYDVQQKEw1MZXQncyBFbmNyeXB0MQswCQYDVQQD
EwJSMzAeFw0yNDAyMjkxMDUwNThaFw0yNDA1MjkxMDUwNTdaMB0xGzAZBgNVBAMT
EnBvc3RhbC5kdmlzaW9uLm9yZzBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABOw1
Lh25bXePjzUvTd+HklIOi9ZLqn6kkP+6xZO0sSOaOIpX2ROIBuzQKBWrZf7n2WyF
T9rvX+Rg+DxGIEQzxaKjggITMIICDzAOBgNVHQ8BAf8EBAMCB4AwHQYDVR0lBBYw
FAYIKwYBBQUHAwEGCCsGAQUFBwMCMAwGA1UdEwEB/wQCMAAwHQYDVR0OBBYEFH5+
OgworAy6eEQhEiS/3iWlbvPmMB8GA1UdIwQYMBaAFBQusxe3WFbLrlAJQOYfr52L
FMLGMFUGCCsGAQUFBwEBBEkwRzAhBggrBgEFBQcwAYYVaHR0cDovL3IzLm8ubGVu
Y3Iub3JnMCIGCCsGAQUFBzAChhZodHRwOi8vcjMuaS5sZW5jci5vcmcvMB0GA1Ud
EQQWMBSCEnBvc3RhbC5kdmlzaW9uLm9yZzATBgNVHSAEDDAKMAgGBmeBDAECATCC
AQMGCisGAQQB1nkCBAIEgfQEgfEA7wB2AEiw42vapkc0D+VqAvqdMOscUgHLVt0s
gdm7v6s52IRzAAABjfS0EpEAAAQDAEcwRQIgBCLFStiEyF/6rqoVly+BYlGlX9qL
xLPLDH5d4qS47jwCIQCLhKthnfZ6igdwVsfoJO0U5YsaThwajJkDUqPI8sKdsgB1
AO7N0GTV2xrOxVy3nbTNE6Iyh0Z8vOzew1FIWUZxH7WbAAABjfS0EpMAAAQDAEYw
RAIgQ97UKX9qAGVfEp30FR/zv0T+jVlROqoJQY5VO7UPs/MCIGJ45A6be1A+E0RI
bbY4S5Xjn8+Wl9WOEvbADn3u7vGlMA0GCSqGSIb3DQEBCwUAA4IBAQBQrEPhnVZ5
eGwt1OomoZMp6Z7PJd6jbuBrgXJAwAI0DNynccmCtQfcEpParYNONP80r4aiTxeL
FRvOgijnTwrKjxX3OvaqKmJsJOfuuNJasmSuW6KK6zgUDM3606EzUhCUQFrvBAIt
e0In1jKhuwwcVaQraEGeKNsvo1rxSetw9JNtBIricLu4tte+DKCH8WAvb5nwTbzO
G63+NtmEE/GLO4C0HAWSbYoEtqcx63nZqQGFbnxgl1bGI1k2DGfQ0v2ADczrs4+1
DV6Bt5MjAX0DX3uF35Hmsmre5bBMBy8ireInAwwb9dyr+V5AxaKUo41nSMJBo4UX
qaXvXfG8MGb5
-----END CERTIFICATE-----
subject=CN = postal.my-domain.com
issuer=C = US, O = Let's Encrypt, CN = R3
---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: ECDSA
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 2886 bytes and written 437 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 256 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
250 AUTH CRAM-MD5 PLAIN LOGIN
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: 003493CA6DE7DCD27E990823955AA3CE2F6F8FE69342902B7E355F752222D4FC
    Session-ID-ctx: 
    Resumption PSK: C471212E918DE31CAE3EB0774DE4B1FF04C31AEFD851444ADFF7197597F8FE42CACC3930EE6FC15C8BB3BBEA3A5198AF
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    0000 - fe d0 9e 0b 59 e8 a6 82-6d 9b 6e 13 24 4f 41 ef   ....Y...m.n.$OA.
    0010 - 33 ef d7 fc 6f 20 d4 45-cf 90 cc d1 0b d8 bc a2   3...o .E........
    0020 - ab c2 f4 32 ad 74 63 72-0c 01 67 ce b1 8b 21 9c   ...2.tcr..g...!.
    0030 - 70 f2 9f f4 a9 0c 91 cc-fb b4 c1 6b db c7 52 5e   p..........k..R^
    0040 - e0 fd b3 1d 31 63 3a 72-7f 37 8f 16 fe 36 1a c8   ....1c:r.7...6..
    0050 - 33 57 a2 00 ff 08 39 93-f4 8d df 16 e2 46 7d 1d   3W....9......F}.
    0060 - 28 5b e9 17 29 61 1d 93-7e 24 22 98 0e 9f c7 65   ([..)a..~$"....e
    0070 - 78 79 17 2e 00 00 91 30-d6 30 06 a2 98 1c 8e 33   xy.....0.0.....3
    0080 - a9 f8 8e 7f 01 66 de a0-fa f7 bf 80 97 96 98 34   .....f.........4
    0090 - 28 9b 0b 50 8c 4d ec 5f-c9 68 d4 30 f9 2f f8 14   (..P.M._.h.0./..
    00a0 - c3 55 fc d1 6c 07 df 75-53 a5 ce d9 57 cc 80 54   .U..l..uS...W..T
    00b0 - 21 b8 ab a1 06 f6 89 31-b4 5d 9b 88 b5 1a e2 04   !......1.]......

    Start Time: 1709211887
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: C84208C84AEDBE91B8A5EDB5D1AB4B7642C8D13F73975EED86BC635E316645BA
    Session-ID-ctx: 
    Resumption PSK: 8C896B81DD53F042647A654A61BABCEEBE5C9B5283C7701179C969339213D8F362E16507DDBD08143D6CCE4F284ACD99
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    0000 - fe d0 9e 0b 59 e8 a6 82-6d 9b 6e 13 24 4f 41 ef   ....Y...m.n.$OA.
    0010 - bb 84 24 de 5a dd f3 2f-67 24 50 a5 2a 12 2f f9   ..$.Z../g$P.*./.
    0020 - 30 1a d4 08 9d e2 a9 29-1a 2b f0 1e 57 7c 61 28   0......).+..W|a(
    0030 - 6b 9a 74 36 fd db d2 ed-ec c5 0f 0d 8b ef a1 b9   k.t6............
    0040 - c8 55 72 70 cf 9d 69 dd-c7 3b f6 ed 58 ab 7d ba   .Urp..i..;..X.}.
    0050 - 4c 1a 7f 65 50 78 d2 96-8f f4 cf 14 0b 3d b6 cf   L..ePx.......=..
    0060 - 8e 89 00 0b 2f 00 82 c3-da d8 4d 1f b5 9d 52 ad   ..../.....M...R.
    0070 - 60 f3 bd 86 bf 54 fa 59-e1 60 3c a9 a2 ab ab a6   `....T.Y.`<.....
    0080 - d8 0b 5d a9 1e d2 e5 2a-e5 3a dc bc 71 4f d4 ed   ..]....*.:..qO..
    0090 - 49 22 d2 ab fd 4c 1e 7b-be 91 09 01 20 ab d6 a3   I"...L.{.... ...
    00a0 - 6e 8c 84 41 71 fd a3 39-1d 58 4d 5b 2d e4 24 19   n..Aq..9.XM[-.$.
    00b0 - 78 7a cc 7f dc 4e f1 e2-fb 7c bf d6 c3 7a 6b 5b   xz...N...|...zk[

    Start Time: 1709211888
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK
```
(the command stops after that - dont know if this is supposed to)

Can you give me further assistance? Thank you very much

**Comment by willpower232** [comment]
> ERROR: An error occurred while attempting to establish an SSL or TLS connection.

this is where you need to see the logs from

You could also try `postal logs smtp` to see what is happening on Postals end.

It seems like everything is working so it is more likely that whatever you're trying to use is perhaps changing the port from 25 to one of the traditionally used ports for TLS connections.

If this is the case then the logs would not change when you are testing out the other connection as nothing is reaching Postal.

**Reply by david-brombacher** [reply]
If i run the command:

`openssl s_client -connect postal.my-domain.com:25`

i get:
```
CONNECTED(00000003)
40075ADB167F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../ssl/record/ssl3_record.c:354:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 5 bytes and written 324 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
``` 
At the same time it says this in `postal logs smtp`:
![image](https://github.com/postalserver/postal/assets/70525131/b95199a0-dea7-4401-bec2-5072845e807d)


I tried to set 
`SSL_version: SSLv3`

in the postal config, but this wont work and will stop TLS from working.

**Reply by adamcooke** [reply, marked as answer]
Postal's SMTP server only supports STARTTLS, it does not support direct TLS connections (SMTPS).

**Reply by david-brombacher** [reply]
I understand now - it is confusion, because i have clients working with configuration like:

[  ] none
[  ] SSL
[X] TLS

so i think it will use STARTTLS (even if written as TLS).

The problem is still, that i have at least 3 different clients not working on v2, that worked on my postal v1.x server. I even made a new postal installation on a different FQDN - same problem. The problematic clients are set to "STARTTLS" - so i dont understand, why they stopped working. Since the command `openssl s_client -starttls smtp -crlf -connect postal.my-domain.com:25` gives a good looking response i believe (result in previous post).

I did run the same command against my postal v1 server, which returned a similar looking result, but i noticed that the old server says TLSv1.2 while the new shows TLSv1.3 - Should i try changing this on the new one?
